yb-voyager/src/srcdb/yugabytedb.go

```diff
@@ -611,74 +611,179 @@ func (yb *YugabyteDB) FilterUnsupportedTables(migrationUUID uuid.UUID, tableList
 }
 
 func (yb *YugabyteDB) FilterEmptyTables(tableList []sqlname.NameTuple) ([]sqlname.NameTuple, []sqlname.NameTuple) {
-	var nonEmptyTableList, emptyTableList []sqlname.NameTuple
-	for _, tableName := range tableList {
-		query := fmt.Sprintf(`SELECT false FROM %s LIMIT 1;`, tableName.ForUserQuery())
-		var empty bool
-		err := yb.db.QueryRow(query).Scan(&empty)
-		if err != nil {
-			if err == sql.ErrNoRows {
-				empty = true
-			} else {
-				utils.ErrExit("error in querying table: %v: %w", tableName, err)
-			}
+	if len(tableList) == 0 {
+		return nil, nil
+	}
+
+	// Build a single UNION ALL query to check all tables at once.
+	// Example query for 3 tables:
+	//   SELECT 0 AS table_idx, EXISTS(SELECT 1 FROM public.products) AS has_rows
+	//   UNION ALL
+	//   SELECT 1 AS table_idx, EXISTS(SELECT 1 FROM public.users) AS has_rows
+	//   UNION ALL
+	//   SELECT 2 AS table_idx, EXISTS(SELECT 1 FROM public.invoices) AS has_rows
+	var unionParts []string
+	for idx, tableName := range tableList {
+		unionParts = append(unionParts,
+			fmt.Sprintf("SELECT %d AS table_idx, EXISTS(SELECT 1 FROM %s) AS has_rows",
+				idx, tableName.ForUserQuery()))
+	}
+	query := strings.Join(unionParts, " UNION ALL ")
+
+	rows, err := yb.db.Query(query)
+	if err != nil {
+		utils.ErrExit("failed to query for empty tables: %w", err)
+	}
+	defer rows.Close()
+
+	tableIsEmpty := make([]bool, len(tableList)) // defaults to false for all tables
+	for rows.Next() {
+		var tableIdx int
+		var hasRows bool
+		if err := rows.Scan(&tableIdx, &hasRows); err != nil {
+			utils.ErrExit("failed to scan row for empty table check: %w", err)
+		}
+		if !hasRows {
+			tableIsEmpty[tableIdx] = true
 		}
-		if !empty {
-			nonEmptyTableList = append(nonEmptyTableList, tableName)
+	}
+
+	if err := rows.Err(); err != nil {
+		utils.ErrExit("failed to iterate rows for empty table check: %w", err)
+	}
+
+	var nonEmptyTableList, emptyTableList []sqlname.NameTuple
+	for i, isEmpty := range tableIsEmpty {
+		if isEmpty {
+			emptyTableList = append(emptyTableList, tableList[i])
 		} else {
-			emptyTableList = append(emptyTableList, tableName)
+			nonEmptyTableList = append(nonEmptyTableList, tableList[i])
 		}
 	}
 	return nonEmptyTableList, emptyTableList
 }
 
-func (yb *YugabyteDB) getTableColumns(tableName sqlname.NameTuple) ([]string, []string, []string, error) {
-	var columns, dataTypes, dataTypesOwner []string
-	sname, tname := tableName.ForCatalogQuery()
-	query := fmt.Sprintf(GET_TABLE_COLUMNS_QUERY_TEMPLATE_PG_AND_YB, tname, sname)
+// tableColumnInfo holds column information for a table
+type tableColumnInfo struct {
+	Columns   []string
+	DataTypes []string
```

Comment:
instead you can have map column->datatype also right?

Reply:
@priyanshi-yb map doesn't ensure the order (its random) thats why i chose this as its safer also we have some logic relying on this.

Do you think it will not matter?
I don't have the exact reason but seems risky to me, as we pass this column list to debezium and pg uses this.

Reply:
https://github.com/yugabyte/yb-voyager/pull/3151#discussion_r2476982735